File: client/src/messagehistory.cpp
#include "messagehistory.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QLocale>
#include <QStandardPaths>
#include <QDebug>

namespace {
const int kMaxEntries = 200;   // a quick-message log, not an unbounded
                                // chat archive -- kept small on purpose.

// English-only, pinned locale (not the device locale) so month names
// don't vary unexpectedly -- same choice made on the Android side.
QString formatTimestamp(qint64 ms)
{
    QLocale en(QLocale::English, QLocale::UnitedStates);
    return en.toString(QDateTime::fromMSecsSinceEpoch(ms), QStringLiteral("MMM d hh:mm AP"));
}
}

MessageHistory::MessageHistory(QObject *parent)
    : QObject(parent)
{
    load();
}

QString MessageHistory::filePath() const
{
    const QString dir = QStandardPaths::writableLocation(QStandardPaths::AppDataLocation);
    QDir().mkpath(dir);          // first run: the directory may not exist yet
    return dir + QStringLiteral("/message_history.json");
}

void MessageHistory::addSent(const QString &id, const QString &peer, const QString &message)
{
    QVariantMap e;
    e.insert(QStringLiteral("id"), id);
    e.insert(QStringLiteral("direction"), QStringLiteral("sent"));
    e.insert(QStringLiteral("peer"), peer);
    e.insert(QStringLiteral("message"), message);
    const qint64 now = QDateTime::currentMSecsSinceEpoch();
    e.insert(QStringLiteral("timestampMs"), now);
    e.insert(QStringLiteral("timestampText"), formatTimestamp(now));
    e.insert(QStringLiteral("status"), QStringLiteral("pending"));
    m_entries.prepend(e);         // newest first
    trimAndPublish();
}

void MessageHistory::addReceived(const QString &id, const QString &peer, const QString &message)
{
    QVariantMap e;
    e.insert(QStringLiteral("id"), id);
    e.insert(QStringLiteral("direction"), QStringLiteral("received"));
    e.insert(QStringLiteral("peer"), peer);
    e.insert(QStringLiteral("message"), message);
    const qint64 now = QDateTime::currentMSecsSinceEpoch();
    e.insert(QStringLiteral("timestampMs"), now);
    e.insert(QStringLiteral("timestampText"), formatTimestamp(now));
    e.insert(QStringLiteral("status"), QStringLiteral("delivered"));  // meaningless
    m_entries.prepend(e);          // for received, but harmless to set
    trimAndPublish();
}

void MessageHistory::markStatus(const QString &id, const QString &status)
{
    for (int i = 0; i < m_entries.size(); ++i) {
        QVariantMap e = m_entries.at(i).toMap();
        if (e.value(QStringLiteral("id")).toString() == id
                && e.value(QStringLiteral("direction")).toString() == QLatin1String("sent")) {
            e.insert(QStringLiteral("status"), status);
            m_entries[i] = e;
            save();
            emit entriesChanged();
            return;
        }
    }
    // No matching SENT row (e.g. a delivered-ack for a message this
    // device never sent) -- nothing to update, not an error.
}

void MessageHistory::clear()
{
    m_entries.clear();
    save();
    emit entriesChanged();
}

void MessageHistory::trimAndPublish()
{
    while (m_entries.size() > kMaxEntries)
        m_entries.removeLast();
    save();
    emit entriesChanged();
}

void MessageHistory::load()
{
    m_entries.clear();
    QFile f(filePath());
    if (!f.exists()) {
        return;                    // first run: nothing to restore, not an error
    }
    if (!f.open(QIODevice::ReadOnly)) {
        qWarning() << "paniccall: could not open message history file:" << f.errorString();
        return;
    }
    const QByteArray raw = f.readAll();
    f.close();
    QJsonParseError err;
    const QJsonDocument doc = QJsonDocument::fromJson(raw, &err);
    if (err.error != QJsonParseError::NoError || !doc.isArray()) {
        // Corrupt/unreadable history must never crash the app; start
        // fresh -- same "warn and continue, never fatal" spirit as the
        // relay's own corrupt-pending-state handling.
        qWarning() << "paniccall: message history file is corrupt, starting fresh:"
                   << err.errorString();
        return;
    }
    const QJsonArray arr = doc.array();
    for (const QJsonValue &v : arr) {
        if (v.isObject())
            m_entries.append(v.toObject().toVariantMap());
    }
}

void MessageHistory::save()
{
    QJsonArray arr;
    for (const QVariant &v : m_entries)
        arr.append(QJsonObject::fromVariantMap(v.toMap()));
    QFile f(filePath());
    if (!f.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
        // Best-effort: losing a history write is far less bad than
        // crashing the call/text feature over it.
        qWarning() << "paniccall: could not write message history file:" << f.errorString();
        return;
    }
    f.write(QJsonDocument(arr).toJson(QJsonDocument::Compact));
    f.close();
}

File: client-android/app/src/main/java/com/mykelsilver/paniccall/MessageHistory.kt
package com.mykelsilver.paniccall

import android.content.Context
import kotlinx.coroutines.flow.MutableStateFlow
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Local, per-device message history: what YOU sent (with delivery status)
 * and what you received. Persisted as a small JSON file in app-internal
 * storage -- deliberately NOT stored on the relay (see docs/PROTOCOL.md:
 * the relay keeps no message content beyond a single in-flight pending
 * message). Survives app restarts; does not survive an uninstall/reinstall,
 * same tradeoff already accepted for the relay's own durability design,
 * and each device only ever sees its own side of the conversation.
 */
class MessageHistory(context: Context) {

    enum class Direction { SENT, RECEIVED }
    enum class Status { PENDING, SENT, QUEUED, DELIVERED }

    data class Entry(
        val id: String,
        val direction: Direction,
        val peer: String,          // the other person's display name
        val message: String,
        val timestampMs: Long,
        val status: Status,
    )

    private val file = File(context.filesDir, "message_history.json")
    private val entries = mutableListOf<Entry>()   // newest first

    val history = MutableStateFlow<List<Entry>>(emptyList())

    init { load() }

    /** Call right after CallEngine.sendText() returns its id. */
    @Synchronized
    fun addSent(id: String, peer: String, message: String) {
        entries.add(0, Entry(id, Direction.SENT, peer, message,
            System.currentTimeMillis(), Status.PENDING))
        trimAndPublish()
    }

    /** Call when CallEngine.textReceived fires. */
    @Synchronized
    fun addReceived(id: String, peer: String, message: String) {
        entries.add(0, Entry(id, Direction.RECEIVED, peer, message,
            System.currentTimeMillis(), Status.DELIVERED))
        trimAndPublish()
    }

    /** Call on text_sent (SENT/QUEUED) and text_delivered (DELIVERED) acks. */
    @Synchronized
    fun markStatus(id: String, status: Status) {
        val idx = entries.indexOfFirst { it.id == id && it.direction == Direction.SENT }
        if (idx < 0) return             // e.g. a delivered-ack for a message
        entries[idx] = entries[idx].copy(status = status)   // this device never sent
        save()
        history.value = entries.toList()
    }

    private fun trimAndPublish() {
        // A quick-message log, not an unbounded chat archive -- keep it
        // small on purpose, matching the feature's own scope.
        while (entries.size > MAX_ENTRIES) entries.removeAt(entries.size - 1)
        save()
        history.value = entries.toList()
    }

    private fun load() {
        entries.clear()
        try {
            if (file.exists()) {
                val arr = JSONArray(file.readText())
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    entries.add(Entry(
                        id = o.optString("id"),
                        direction = Direction.valueOf(o.optString("direction", "SENT")),
                        peer = o.optString("peer"),
                        message = o.optString("message"),
                        timestampMs = o.optLong("timestampMs"),
                        status = Status.valueOf(o.optString("status", "SENT")),
                    ))
                }
            }
        } catch (e: Exception) {
            // Corrupt/unreadable history must never crash the app; start
            // fresh -- same "warn and continue, never fatal" spirit as the
            // relay's own corrupt-pending-state handling.
            entries.clear()
        }
        history.value = entries.toList()
    }

    private fun save() {
        try {
            val arr = JSONArray()
            for (e in entries) {
                arr.put(JSONObject()
                    .put("id", e.id)
                    .put("direction", e.direction.name)
                    .put("peer", e.peer)
                    .put("message", e.message)
                    .put("timestampMs", e.timestampMs)
                    .put("status", e.status.name))
            }
            file.writeText(arr.toString())
        } catch (e: Exception) {
            // Best-effort: losing a history write is far less bad than
            // crashing the call/text feature over it.
        }
    }

    companion object {
        private const val MAX_ENTRIES = 200
    }
}
